packages/common/src/hooks/useImageSize2.ts

```diff
@@ -0,0 +1,134 @@
+
+  const fetchWithFallback = useCallback(
+    async (url: string) => {
+      const mirrors = [...(artwork?.mirrors ?? [])]
+      let currentUrl = url
+
+      while (mirrors.length > 0) {
+        try {
+          await preloadImage(currentUrl)
+          return currentUrl
+        } catch {
+          const nextMirror = mirrors.shift()
+          if (!nextMirror) throw new Error('No mirror found')
+          const currentHost = new URL(currentUrl).hostname
+          const nextHost = new URL(nextMirror).hostname
+          const newUrl = currentUrl.replace(currentHost, nextHost)
+          currentUrl = newUrl
+        }
+      }
+
+      throw new Error('Failed to fetch image from all mirrors')
+    },
+    [artwork?.mirrors]
+  )
+
+  const resolveImageUrl = useCallback(async () => {
+    if (!artwork) {
+      console.debug(`useImageSize: no artwork, loading`)
+      return
+    }
+    const targetUrl = artwork[targetSize]
+    if (!targetUrl) {
+      console.debug(`useImageSize: no target url for ${targetSize}`)
+      setImageUrl(defaultImage)
+      return
+    }
+
+    if (IMAGE_CACHE.has(targetUrl)) {
+      console.debug(`useImageSize: cache hit ${targetUrl}`)
+      setImageUrl(targetUrl)
+      return
+    }
+
+    // Check for smaller size in the cache
+    // If found, set the image url and preload the actual target url
+    const smallerSize = Object.keys(artwork).find(
+      (size) =>
+        parseInt(size) < parseInt(targetSize) &&
+        artwork[size as SquareSizes] &&
+        IMAGE_CACHE.has(artwork[size as SquareSizes]!)
+    ) as SquareSizes | undefined
+
+    // Check for larger size
+    // If found, set the image url to the larger size and return
+    const largerSize = Object.keys(artwork).find(
+      (size) =>
+        parseInt(size) > parseInt(targetSize) &&
+        artwork[size as SquareSizes] &&
+        IMAGE_CACHE.has(artwork[size as SquareSizes]!)
+    ) as SquareSizes | undefined
+
+    if (largerSize) {
+      console.debug(
+        `useImageSize: cache miss, found larger ${targetUrl} ${largerSize}`
+      )
+      setImageUrl(artwork[largerSize] ?? null)
+      return
+    }
+
+    if (smallerSize) {
+      console.debug(
+        `useImageSize: cache miss, found smaller ${targetUrl} ${smallerSize}`
+      )
+      setImageUrl(artwork[smallerSize] ?? null)
+      const finalUrl = await fetchWithFallback(targetUrl)
+      IMAGE_CACHE.add(finalUrl)
+      setImageUrl(finalUrl)
+      return
+    }
+
+    // Fetch image with fallback mirrors
+    try {
+      console.debug(`useImageSize: cache miss fetch original ${targetUrl}`)
+      const finalUrl = await fetchWithFallback(targetUrl)
+      IMAGE_CACHE.add(finalUrl)
+      setImageUrl(finalUrl)
+    } catch {
+      console.error(`Unable to load image ${targetUrl} after retries`)
+    }
+  }, [artwork, targetSize, fetchWithFallback, defaultImage])
+
+  useEffect(() => {
+    resolveImageUrl()
+  }, [resolveImageUrl])
+
+  return imageUrl ?? undefined
```

Comment:
Why the coalesce to undefined? Backwards-compatibility? Might need a comment here.

Reply:
nah good call the hook had null instead of undefined. this was a hack while i was writing it. cleaned up